compiler/rustc_type_ir/src/fast_reject.rs

```diff
@@ -173,27 +171,46 @@ impl<DefId> SimplifiedType<DefId> {
     }
 }
 
-/// Given generic arguments from an obligation and an impl,
-/// could these two be unified after replacing parameters in the
-/// the impl with inference variables.
+/// Given generic arguments, could they be unified after
+/// replacing parameters with inference variables or placeholders.
+/// This behavior is toggled using the const generics.
 ///
-/// For obligations, parameters won't be replaced by inference
-/// variables and only unify with themselves. We treat them
-/// the same way we treat placeholders.
+/// We use this to quickly reject impl/wc candidates without needing
+/// to instantiate generic arguments/having to enter a probe.
 ///
 /// We also use this function during coherence. For coherence the
 /// impls only have to overlap for some value, so we treat parameters
-/// on both sides like inference variables. This behavior is toggled
-/// using the `treat_obligation_params` field.
+/// on both sides like inference variables.
 #[derive(Debug, Clone, Copy)]
-pub struct DeepRejectCtxt<I: Interner> {
-    treat_obligation_params: TreatParams,
+pub struct DeepRejectCtxt<I: Interner, const TREAT_LHS_PARAMS: bool, const TREAT_RHS_PARAMS: bool> {
     _interner: PhantomData<I>,
 }
 
-impl<I: Interner> DeepRejectCtxt<I> {
-    pub fn new(_interner: I, treat_obligation_params: TreatParams) -> Self {
-        DeepRejectCtxt { treat_obligation_params, _interner: PhantomData }
+impl TreatParams {
+    pub const fn into_bool(&self) -> bool {
+        match *self {
+            TreatParams::InstantiateWithInfer => true,
+            TreatParams::AsRigid => false,
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! new_reject_ctxt {
+    ($interner:expr, $lhs:ident, $rhs:ident) => {
+        $crate::fast_reject::DeepRejectCtxt::<
+            _,
+            { $crate::fast_reject::TreatParams::$lhs.into_bool() },
+            { $crate::fast_reject::TreatParams::$rhs.into_bool() },
+        >::new($interner)
+    };
+}
+
+impl<I: Interner, const TREAT_LHS_PARAMS: bool, const TREAT_RHS_PARAMS: bool>
```

Comment:
```suggestion
impl<I: Interner, const INSTANTIATE_LHS_WITH_INFER: bool, const INSTANTIATE_RHS_WITH_INFER: bool>
```
so that it's clear what `true` and `false` represent.

Instead of using a macro, it might be better to add constructors to `DeepRejectCtxt`, e.g.
```rust
impl<I: Interner> DeepRejectCtxt<I, false, false> {
    /// Treat parameters in both the lhs and the rhs as rigid.
    fn rigid_rigid(_interner: I) -> DeepRejectCtxt<I, false, false> {
        DeepRejectCtxt { _interner: PhantomData }
    }
}
```